internal/remediation/override.go

```diff
@@ -279,6 +279,58 @@ func getVersionsGreater(ctx context.Context, cl client.DependencyClient, vk reso
 	return versions[offset:], nil
 }
 
+// comparisonFunctionWithWorkarounds returns a version comparison function with special behaviour for specific packages,
+// producing more desirable ordering using non-standard comparison.
+func comparisonFunctionWithWorkarounds(vk resolve.VersionKey) func(resolve.Version, resolve.Version) int {
+	sys := vk.Semver()
+
+	if vk.System == resolve.Maven && vk.Name == "com.google.guava:guava" {
+		// com.google.guava:guava has 'flavors' with versions ending with -jre or -android.
+		// https://github.com/google/guava/wiki/ReleasePolicy#flavors
+		// To preserve the flavor in updates, we make the opposite flavor considered the earliest versions.
+
+		// Old versions have '22.0' and '22.0-android', and even older version don't have any flavors.
+		// Only check for the android flavor, and assume its jre otherwise.
+		wantAndroid := strings.HasSuffix(vk.Version, "-android")
+		return func(a, b resolve.Version) int {
+			aIsAndroid := strings.HasSuffix(a.Version, "-android")
+			bIsAndroid := strings.HasSuffix(b.Version, "-android")
+
+			if aIsAndroid == bIsAndroid {
+				return sys.Compare(a.Version, b.Version)
+			}
+
+			if aIsAndroid == wantAndroid {
+				return 1
+			}
+
+			return -1
+		}
+	}
+
+	if vk.System == resolve.Maven && strings.HasPrefix(vk.Name, "commons-") {
+		// Old versions of apache commons- libraries (commons-io:commons-io, commons-math:commons-math, etc.)
+		// used date-based versions (e.g. 20040118.003354), which naturally sort after the more recent semver versions.
+		// We manually force the date versions to come before the others to prevent downgrades.
+		return func(a, b resolve.Version) int {
+			aCal := strings.HasPrefix(a.Version, "200")
```

Comment:
I think it's better to check if the first 8 numbers represent a date. https://pkg.go.dev/time#Parse

Reply:
It seems [pretty inefficient](https://cs.opensource.google/go/go/+/refs/tags/go1.23.0:src/time/format.go;l=1044) to keep calling `Time.Parse("20060102", a.Version[:8])` repeatedly, especially since we don't actually need to use the parsed year/month/day.

I've taken a quick look at the `commons-*` packages on Maven central, and the date versions seem to only have been used from the years 2002-2005 (which seems to be up to the release of Maven version 2).
I don't think we'll see any of these packages versioned like this post 2010, and I also doubt we'd ever see the major versions of these packages reach up to `200.0.0`.

Reply:
I think we can filter out most of the versions by checking if the length is greater than 8, or maybe even implement our own simplified checker. if all the versions are from year 2002-2005, add a comment about this specialty.

Reply:
I've added a comment to the code justify checking only the first 3 characters.

Checking it properly would ultimately come down to doing a similar check, and I'd rather not over-engineer it for a case that doesn't exist.